batch.py

```diff
@@ -216,17 +506,186 @@ def make_gene_loc_dict(file) -> dict[str, dict]:
     return gene_dict
 
 
-def remove_sc_outliers(df, outliers=['966_967', '88_88']):
+# can probably be merged with below
+def extract_genes(gene_list, expression_tsv_path) -> list[str]:
+    """
+    Takes a list of all genes and subsets to only those
+    present in the expression file of interest
+    """
+    expression_df = pd.read_csv(AnyPath(expression_tsv_path), sep='\t')
+    expression_df = filter_lowly_expressed_genes(expression_df)
+    gene_ids = set(list(expression_df.columns.values)[1:])
+    genes = set(gene_list).intersection(gene_ids)
+
+    logging.info(f'Total genes to run: {len(list(sorted(genes)))}')
+
+    return list(sorted(genes))
+
+
+# copied from https://github.com/populationgenomics/tob-wgs/blob/main/scripts/eqtl_hail_batch/launch_eqtl_spearman.py
+# generalised to specify min pct samples as input
+def filter_lowly_expressed_genes(expression_df, min_pct=10):
+    """Remove genes with low expression in all samples
+    Input:
+    expression_df: a data frame with samples as rows and genes as columns,
+    containing normalised expression values (i.e., the average number of molecules
+    for each gene detected in each person).
+    Returns:
+    A filtered version of the input data frame, after removing columns (genes)
+    with 0 values in more than {min_pct}% of the rows (samples) - 10 by default.
+    """
+    # Remove genes with 0 expression in all samples
+    expression_df = expression_df.loc[:, (expression_df != 0).any(axis=0)]
+    genes_not_equal_zero = expression_df.iloc[:, 1:].values != 0
+    n_expr_over_zero = pd.DataFrame(genes_not_equal_zero.sum(axis=0))
+    percent_expr_over_zero = (n_expr_over_zero / len(expression_df.index)) * 100
+    percent_expr_over_zero.index = expression_df.columns[1:]
+
+    # Filter genes with less than 10 percent individuals with non-zero expression
+    atleastNpercent = percent_expr_over_zero[(percent_expr_over_zero > min_pct)[0]]
+    sample_ids = expression_df['sampleid']
+    expression_df = expression_df[atleastNpercent.index]
+    expression_df.insert(loc=0, column='sampleid', value=sample_ids)
+
+    return expression_df
+
+
+def remove_sc_outliers(df, outliers=None):
     """
     Remove outlier samples, as identified by single-cell analysis
     """
+    if outliers is None:
+        outliers = ['966_967', '88_88']
+    else:
+        outliers = outliers.extend(['966_967', '88_88'])
     df = df[-df['OneK1K_ID'].isin(outliers)]
 
     return df
 
 
+def qvalue(pv, m=None, verbose=False, lowmem=False, pi0=None):
```

Comment:
Why do we need to copy the source instead of using the `qvalue` module?

Reply:
good question, I believe it wasn't up-to-date, and would not work in python 3 (definitely used ```xrange```, for example) - could fork instead / or submit an issue I suppose?

Reply:
That's really odd, because @KatalinaBobowik is using it in the code below, or is that a different implementation?
https://github.com/populationgenomics/tob-wgs/blob/88cce4d2d82f29388b5cdce3ef6715e0a918218b/scripts/eqtl_hail_batch/launch_eqtl_spearman.py#L803

Reply:
ah interesting, yes it's a different implementation.

Kat's using qvalue from multipy.fdr (https://github.com/puolival/multipy)

I (was attempting to) use https://github.com/nfusi/qvalue

they're probably equivalent, i was just using one I am familiar with :)